[feat] 테스트 코드 작성

epitone opened the pull request:
## 작업사항
1. 테스트 커버리지 볼 수 있도록 적용
2. jest, supertest 설치
3. 유저 회원가입 성공, 실패에 대한 경우의 수 테스트
4. faker 라이브러리 설치해서, 임의의 이메일, 이름을 사용하도록 로직 설정


## 관계된 이슈, PR

epitone committed:
[feat]:  사용자 회원가입 테스트 코드 작성
1. 테스트 커버리지 볼 수 있도록 적용
2. jest, supertest 설치
3. 유저 회원가입 성공, 실패에 대한 경우의 수 테스트
4. faker 라이브러리 설치해서, 임의의 이메일, 이름을 사용하도록 로직 설정

epitone committed:
[fix]: gitignore test 폴더 주석처리

1. 깃허브에 test 폴더 들어가게끔 수정

epitone commented:
@Makkiato @SibaDoge1 

일단 작업하는 부분들 바로 리뷰하실 수 있도록 올려두겠습니다. 

Mock 함수에 대한 이해가 부족해서, faker 라이브러리를 활용해서, 테스트코드를 구성했습니다. 
faker 라이브러리에 대한 정보는 여기서 확인 가능합니다. https://www.npmjs.com/package/faker

epitone committed:
[feat]: 사용자 로그인 테스트 코드 작성

1. 유저 로그인 성공 테스트 코드 작성
2. 유저 로그인 입력값 누락 테스트 코드 작성
3. 유저 로그인 존재하지 않는 이메일 테스트 코드 작성
4. 유저 로그인 비밀번호 불일치 경우 테스트 코드 작성

Makkiato commented on src/test/User/signin.test.js:
```diff
@@ -0,0 +1,65 @@
+const app = require('../../app');
```
실제 express를 가동시키지만

Makkiato commented on src/test/User/signin.test.js:
```diff
@@ -0,0 +1,65 @@
+const routes = require('../../globals/routes');
+const faker = require('faker');
+
+const testClient = supertest(app);
```
callback이나 promise 없이 테스트를 시작시켜버림.

그래서 express 완전히 준비되기전에 요청이 들어가게 되고 정상적이지 않은 flow 발생

차후 app.listen이 도달하는 과정도 검토해보아야 함

epitone replied:
@Makkiato 정말 좋은 의견이십니다. 그래서 말씀대로 일단 임시적으로 로컬에서 테스트할 수 있게끔 설정했습니다.

Makkiato replied:
이런 식으로 수정하면 현재 시스템에서도 가능은 할것으로 예상.

수정부분은 "!" 로 표시

/src/models/index.js : 42~55

! const connectDB = new Promise( async (resolve, reject) => {
  try {
    await modules.sequelize.authenticate();
    await modules.sequelize.sync({ alter: true });
    logger.log('MySQL connected ...');
!   resolve(THIS_DB_CONTROLL_MODULE)
  } catch (err) {
!    //console.error(err.message);
!    //process.exit(1);
!   reject(err)
  }
};)

! module.export.connectDB = connectDB


/src/bin/www.js : 15 ~ 17

! var dbManager = require("PATH TO /src/models/index.js")
! dbManager.connectDB
!   .then(THIS_DB_CONTROLL_MODULE => {
         server.listen(port, function () {
             logger.log(`🔥Server on! http://localhost:${port}`)
         });
!   }).catch(err => {console.log(err)})


그리고 사실 이 부분은 굳이 이 테스트 작업 때문만이 아니어도, 필요한 작업.
애초에 서버는 DB 연결이 되어있는지도 모르는데 request 받겠다고 listen 시작하는것은 무리가 있는듯

epitone replied:
@Makkiato 오 한번 적용시켜보겠습니다. 감사합니다!!

epitone committed:
[Feat] 메뉴 생성 테스트 코드 작성 (#10)

1. 토큰을 발급받아서, 헤더에 토큰을 넣을 수 있게끔 설정
2. 테스트 파일의 host를 localhost로 변경 후, 테스트 클라이언트 설정

To Do List
1. 토큰 발급하는 로직의 경우 중복되는 로직이기 때문에, 코드의 중복성을 피해야함
2. 테스트 커버리지를 올리려면, 토큰 만료, 사용불가 토큰 등의 테스트 경우의 수를 추가해야 함

Makkiato commented on src/test/Menu/createMenu.test.js:
```diff
@@ -0,0 +1,67 @@
+describe('토큰 생성하기', () => {
+  test('토큰 생성 성공', async () => {
+    const res = await testClient
+      .post(`${routes.user}${routes.signin}`)
```
`${routes.user}${routes.signin}`
이 표현은 보면 볼수록 적응하기 힘듬 ctrl키 누르고 툴팁을 띄워도 레퍼런스를 하나 더 올라가야 실제 상수값이 보이고 있으니,

아니면 적어도 javadoc 주석처리라도 해야 enumeration이 과해서 가독성이 떨어지는 일은 방지 할수 있을듯

epitone replied:
@Makkiato 좋은 의견 감사합니다..! 다음 부분에는 적용해보겠습니다!

Makkiato commented on src/test/User/signup.test.js:
```diff
@@ -1,11 +1,16 @@
+const dotenv = require('dotenv');
+dotenv.config();
+
+const host = `http://localhost:${process.env.PORT}`
```
환경변수로 표현하려면 host부분도 환경변수로 같이

환경변수에 port만 있는거 보고 있지도 않은 host 찾을수도

epitone committed:
[feat]: 메뉴 수정 테스트 코드 작성 (#11)

1. 토큰을 발급받아서, 헤더에 토큰을 넣을 수 있게끔 설정
2. 테스트에서 메뉴 생성하고, 삭제하는 로직 추가해서 menuId 동적 제어

epitone committed:
[feat]: 메뉴 삭제 테스트 코드 작성 (#12)

1. 토큰을 발급받아서, 헤더에 토큰을 넣을 수 있게끔 설정

epitone committed:
[feat]: 메뉴 조회 테스트 코드 작성 (#13)

1. 토큰을 발급받아서, 헤더에 토큰을 넣을 수 있게끔 설정

epitone committed:
[feat]: 메뉴 전체 조회 테스트 코드 작성 (#17)

epitone committed:
[fix]: 메뉴 삭제하기 테스트 코드 수정

1. menuId의 경우 99999에서 동적으로 변경되게끔 설정

epitone committed:
[feat]: 태그 생성 테스트 코드 추가 (#18)

epitone committed:
[feat]: 태그 수정 테스트 코드 작성 (#19)

epitone committed:
[feat]: 태그 삭제 테스트 코드 작성 (#21)

epitone committed:
[feat]: 태그 조회 테스트 코드 작성 (#26)
